src/lib/operators/join_hash/join_hash_steps.hpp

```diff
@@ -42,15 +47,68 @@ template <typename T>
-// as n because in many cases, we join on primary key attributes where by definition we have only one match on the
-// smaller side.
-using SmallPosList = boost::container::small_vector<RowID, 1>;
+// Stores the mapping from HashedType to positions. Conceptually, this is similar to an (unordered_)multimap, but it
+// has some optimizations for the performance-critical probe() method. Instead of storing the matches directly in the
+// hashmap (think map<HashedType, PosList>), we store an offset. This keeps the hashmap small and makes it easier to
+// cache.
+template <typename HashedType>
+class PosHashTable {
+  // In case we consider runtime to be more relevant, the flat hash map performs better (measured to be mostly on par
+  // with bytell hash map and in some cases up to 5% faster) but is significantly larger than the bytell hash map.
+  //
+  // The hash table stores the relative offset of a SmallPosList (see below) in the vector of SmallPosLists. The range
+  // of the offset does not limit the number of rows in the partition but the number of distinct values. If we end up
+  // with a partition that has more values, the partitioning algorithm is at fault.
+  using Offset = uint32_t;
+  using HashTable = ska::bytell_hash_map<HashedType, Offset>;
+
+  // The small_vector holds the first n values in local storage and only resorts to heap storage after that. 1 is chosen
+  // as n because in many cases, we join on primary key attributes where by definition we have only one match on the
+  // smaller side.
+  using SmallPosList = boost::container::small_vector<RowID, 1>;
+
+ public:
+  explicit PosHashTable(JoinHashBuildMode mode, size_t max_size)
+      : _hash_table(static_cast<size_t>(max_size * 1.2)), _pos_lists(max_size), _mode(mode) {
+    // slightly oversize (x 1.2) the hash table to avoid unnecessary rebuilds
+  }
 
-// In case we consider runtime to be more relevant, the flat hash map performs better (measured to be mostly on par
-// with bytell hash map and in some cases up to 5% faster) but is significantly larger than the bytell hash map.
-template <typename T>
-using HashTable = ska::bytell_hash_map<T, SmallPosList>;
+  // For a value seen on the build side, add its row_id to the table
+  template <typename InputType>
+  void emplace(const InputType& value, RowID row_id) {
+    const auto casted_value = static_cast<HashedType>(value);
+    const auto it = _hash_table.find(casted_value);
+    if (it != _hash_table.end()) {
+      auto& pos_list = _pos_lists[it->second];
+      if (_mode == JoinHashBuildMode::AllPositions) {
+        pos_list.emplace_back(row_id);
+      }
+    } else {
+      auto& pos_list = _pos_lists[_hash_table.size()];
+      pos_list.push_back(row_id);
+      _hash_table.emplace(casted_value, _hash_table.size());
+      Assert(_hash_table.size() < std::numeric_limits<Offset>::max(), "Hash table too big for offset");
```

Comment:
If that really happens, we might need to scrap everything and re-partition. Honestly, I am not sure if we should worry about that too much. We are far away from joining 4B distinct values.

In case this ever comes up, it would be easy to either increase the size to `uint64_t` (sacrificing some performance) or template the type dependent on the input data.

Reply:
Fixed (for now) by trying to ensure that each partition has an upper bound size limit.